File: 3. Largest prime factor.R
# Problem 3

start_time <- Sys.time()

number <- 600851475143

# get factors
factors <- vector()
if (number %% 2 == 0){ # we treat 2 as a separate case because it is the only even prime factor
  number <- number/2
  factors <- 2
  while(number %% 2 == 0){
    number <- number/2
  }
}

candidate <- 3
while(number > 1){
  if(number %% candidate == 0){
    number <- number/candidate
    factors <- c(factors, candidate)
    if(sum(candidate %% factors[factors <= sqrt(candidate)] == 0) == 0) largest_prime_factor <- candidate # all composite numbers have a factor <= sqrt(number)
    while(number %% candidate == 0){
      number <- number/candidate
    }
  }
  candidate <- candidate + 2
}

end_time <- Sys.time()
end_time - start_time

File: 3. Largest prime factor.py
# Problem 3

number = 600851475143

def check_prime(candidate,primes):
    for prime in primes:
        if candidate % prime == 0:
            return False
    return True

def prime_factors(number):
    factors = []
    while number % 2 == 0: #treat 2 & 3 as separate cases
        number = number/2
        factors.append(2)
    while number % 3 == 0:
        number = number/3
        factors.append(3)
    i = 1
    while number > 1:
        candidates = [6*i-1,6*i+1] #all primes > 3 can be written in that form
        for candidate in candidates:
            if check_prime(candidate, factors): 
                while number % candidate == 0:
                    number = number/candidate
                    factors.append(candidate)
        i += 1
    return factors
        
prime_fact = prime_factors(number)
